Use package names

mmmries opened the pull request:
Going along with #35 this introduces one more small change to support package names when loading definitions from files.

__Note: this PR contains the changes from #35 as well so use [this link](https://github.com/mmmries/exprotobuf/compare/load_multiple_protobuf_files...use_package_names) to see a better diff__

I've added this as a new flag so that it does not break backwards compatibility for existing library users. This does not yet give me the naming convention that I would like to use for my library, but I could live with having an extra level of namespacing for now.

/cc @DylanGriffith @bitwalker

bitwalker commented:
@mmmries Have you tried the `use` outside of a module to see if you get what you want? It should work the way you expect.

bitwalker commented:
Looks good though :)

mmmries commented:
I tried to do "use" outside a module but we can create @module attributes
outside of a module. I think I can get the namespacing to work by just
passing in my own namespace of :"Elixir"
On Dec 1, 2015 9:02 AM, "Paul Schoenfelder" <notifications@github.com>
wrote:

> Looks good though :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exprotobuf/pull/36#issuecomment-161013567>.
>

bitwalker commented:
@mmmries Assuming `inject` is false, it shouldn't be creating any attributes outside of the generated module, though I'm looking at the code in master right now, not your changes, so perhaps that's no longer true. I was testing a variety of things just the other day and I was sure I did exactly that (nothing in the file except a `use` statement).

mmmries commented:
I added a new feature to this branch where you can pass an explicit `namespace` when calling `use Protobuf`. This allowed me to request a namespace of `:"Elixir"` when generating protobuf modules without changing any of the behavior for `inject` or any of the naming conventions we have in place.

I also added a README section about these two features for customizing names.

mmmries commented on lib/exprotobuf.ex:
```diff
@@ -8,44 +8,44 @@ defmodule Protobuf do
   alias Protobuf.Utils
 
   defmacro __using__(opts) do
-    namespace = __CALLER__.module
```
This section of code could definitely use a refactoring. For now I was able to keep the diff fairly small, I'm just doing two nested `case` macros instead of a single `case` macro, but I think I would like to come back to cleanup how we are building the `%Config{}` struct here.

mmmries commented:
I confirmed last night that I can use this branch to compiled my entire library of protobuf definitions like this:

```elixir
defmodule Atlas do
  use Protobuf, from: Path.wildcard("definitions/**/*.proto"),
                use_package_names: true,
                namespace: :"Elixir"
end
```

Then I can encode/decode messages with nested messages and all the other fanciness! :tada: :heart: :heart_eyes: :tada: :heart: :heart_eyes: :tada: :heart: :heart_eyes: :tada: :heart: :heart_eyes: 

No change to `inject` was required. Thanks @DyllanGriffith and @bitwalker for all the great feedback on these PRs. I think between #35 and #36 we are ready to release an update.

mmmries commented:
@bitwalker @DylanGriffith just wondering if either of you have had a chance to check this out?

DylanGriffith commented:
@mmmries I really like how that API turned out. I will have a go at compiling my libraries with this some time in the next 24 hours. Thanks heaps for putting in all this work.

mmmries commented:
Sorry to be annoying here, but I don't want to lost momentum on this project. I think this PR (which includes the changes from #35) is a good cut at `1.0.0`. Any chance someone can double-check that?

bitwalker commented:
@mmmries Sorry, it's been a crazy week! I'm comfortable with merging this and pushing a 1.0.0-rc1 release if that sounds good to you!

mmmries commented:
`rc1` is probably a good idea to give people some time to try it out.

bitwalker commented:
1.0.0-rc1 has been pushed to hex :)

DylanGriffith commented:
@mmmries , @bitwalker : I've just tested this on my end. It compiles all our stuff and I really like the new API. Good work guys!

bitwalker commented:
All credit is due to @mmmries, he's done the hard bit :), I'm just glad to get it out there in a release!
